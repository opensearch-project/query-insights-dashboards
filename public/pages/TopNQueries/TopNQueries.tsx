/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useCallback, useEffect, useState } from 'react';
import { Redirect, Route, Switch, useHistory, useLocation } from 'react-router-dom';
import { EuiTab, EuiTabs, EuiTitle, EuiSpacer } from '@elastic/eui';
import { CoreStart } from 'opensearch-dashboards/public';
import QueryInsights from '../QueryInsights/QueryInsights';
import Configuration from '../Configuration/Configuration';
import QueryDetails from '../QueryDetails/QueryDetails';
import { SearchQueryRecord } from '../../../types/types';
import { QueryGroupDetails } from '../QueryGroupDetails/QueryGroupDetails';
import { QueryInsightsDashboardsPluginStartDependencies } from '../../types';
import { PageHeader } from '../../components/PageHeader';
import {
  DEFAULT_TIME_UNIT,
  DEFAULT_TOP_N_SIZE,
  DEFAULT_WINDOW_SIZE,
  MetricType,
} from '../Utils/Constants';

import { MetricSettingsResponse } from '../../types';
import { getTimeAndUnitFromString } from '../Utils/MetricUtils';
import { parseDateString } from '../Utils/DateUtils';

export const QUERY_INSIGHTS = '/queryInsights';
export const CONFIGURATION = '/configuration';

export interface MetricSettings {
  isEnabled: boolean;
  currTopN: string;
  currWindowSize: string;
  currTimeUnit: string;
}

export interface GroupBySettings {
  groupBy: string;
}

const TopNQueries = ({
  core,
  depsStart,
  initialStart = 'now-1d',
  initialEnd = 'now',
}: {
  core: CoreStart;
  depsStart: QueryInsightsDashboardsPluginStartDependencies;
  initialStart?: string;
  initialEnd?: string;
}) => {
  const history = useHistory();
  const location = useLocation();
  const [loading, setLoading] = useState(false);
  const [currStart, setStart] = useState(initialStart);
  const [currEnd, setEnd] = useState(initialEnd);
  const [recentlyUsedRanges, setRecentlyUsedRanges] = useState([
    { start: currStart, end: currEnd },
  ]);
  const [latencySettings, setLatencySettings] = useState<MetricSettings>({
    isEnabled: false,
    currTopN: DEFAULT_TOP_N_SIZE,
    currWindowSize: DEFAULT_WINDOW_SIZE,
    currTimeUnit: DEFAULT_TIME_UNIT,
  });

  const [cpuSettings, setCpuSettings] = useState<MetricSettings>({
    isEnabled: false,
    currTopN: DEFAULT_TOP_N_SIZE,
    currWindowSize: DEFAULT_WINDOW_SIZE,
    currTimeUnit: DEFAULT_TIME_UNIT,
  });

  const [memorySettings, setMemorySettings] = useState<MetricSettings>({
    isEnabled: false,
    currTopN: DEFAULT_TOP_N_SIZE,
    currWindowSize: DEFAULT_WINDOW_SIZE,
    currTimeUnit: DEFAULT_TIME_UNIT,
  });

  const [groupBySettings, setGroupBySettings] = useState<GroupBySettings>({ groupBy: 'none' });

  const setMetricSettings = (metricType: string, updates: Partial<MetricSettings>) => {
    switch (metricType) {
      case MetricType.LATENCY:
        setLatencySettings((prevSettings) => ({ ...prevSettings, ...updates }));
        break;
      case MetricType.CPU:
        setCpuSettings((prevSettings) => ({ ...prevSettings, ...updates }));
        break;
      case MetricType.MEMORY:
        setMemorySettings((prevSettings) => ({ ...prevSettings, ...updates }));
        break;
    }
  };

  const [queries, setQueries] = useState<SearchQueryRecord[]>([]);

  const tabs: Array<{ id: string; name: string; route: string }> = [
    {
      id: 'topNQueries',
      name: 'Top N queries',
      route: QUERY_INSIGHTS,
    },
    {
      id: 'configuration',
      name: 'Configuration',
      route: CONFIGURATION,
    },
  ];

  const onSelectedTabChanged = (route: string) => {
    const { pathname: currPathname } = location;
    if (!currPathname.includes(route)) {
      history.push(route);
    }
  };

  const renderTab = (tab: { route: string; id: string; name: string }) => (
    <EuiTab
      onClick={() => onSelectedTabChanged(tab.route)}
      isSelected={location.pathname.includes(tab.route)}
      key={tab.id}
    >
      {tab.name}
    </EuiTab>
  );

  const retrieveQueries = useCallback(
    async (start: string, end: string) => {
      const nullResponse = { response: { top_queries: [] } };
      const params = {
        query: {
          from: parseDateString(start),
          to: parseDateString(end),
        },
      };
      const fetchMetric = async (endpoint: string) => {
        try {
          // TODO: #13 refactor the interface definitions for requests and responses
          const response: { response: { top_queries: SearchQueryRecord[] } } = await core.http.get(
            endpoint,
            params
          );
          return {
            response: {
              top_queries: Array.isArray(response?.response?.top_queries)
                ? response.response.top_queries
                : [],
            },
          };
        } catch {
          return nullResponse;
        }
      };
      try {
        setLoading(true);
        const respLatency = latencySettings.isEnabled
          ? await fetchMetric('/api/top_queries/latency')
          : nullResponse;
        const respCpu = cpuSettings.isEnabled
          ? await fetchMetric('/api/top_queries/cpu')
          : nullResponse;
        const respMemory = memorySettings.isEnabled
          ? await fetchMetric('/api/top_queries/memory')
          : nullResponse;
        const newQueries = [
          ...respLatency.response.top_queries,
          ...respCpu.response.top_queries,
          ...respMemory.response.top_queries,
        ];
        const noDuplicates: SearchQueryRecord[] = Array.from(
          new Set(newQueries.map((item) => JSON.stringify(item)))
        ).map((item) => JSON.parse(item));
        setQueries(noDuplicates);
      } catch (error) {
        console.error('Error retrieving queries:', error);
      } finally {
        setLoading(false);
      }
    },
    [latencySettings, cpuSettings, memorySettings, core]
  );

  const retrieveConfigInfo = useCallback(
    async (
      get: boolean,
      enabled: boolean = false,
      metric: string = '',
      newTopN: string = '',
      newWindowSize: string = '',
      newTimeUnit: string = '',
      newGroupBy: string = ''
    ) => {
      if (get) {
        try {
          // Helper to get merged settings with transient overwriting persistent
          const getMergedMetricSettings = (
            persistent: MetricSettingsResponse | undefined,
            transient: MetricSettingsResponse | undefined
          ): MetricSettingsResponse => {
            if (transient !== undefined) {
              return transient;
            }
            return {
              ...persistent,
            };
          };

          const getMergedGroupBySettings = (
            persistent: string | undefined,
            transient: string | undefined
          ) => {
            return transient ?? persistent;
          };

          const resp = await core.http.get('/api/settings');
          const persistentSettings = resp?.response?.persistent?.search?.insights?.top_queries;
          const transientSettings = resp?.response?.transient?.search?.insights?.top_queries;
          const metrics = [
            {
              metricType: MetricType.LATENCY,
              metricSetting: getMergedMetricSettings(
                persistentSettings?.latency,
                transientSettings?.latency
              ),
            },
            {
              metricType: MetricType.CPU,
              metricSetting: getMergedMetricSettings(
                persistentSettings?.cpu,
                transientSettings?.cpu
              ),
            },
            {
              metricType: MetricType.MEMORY,
              metricSetting: getMergedMetricSettings(
                persistentSettings?.memory,
                transientSettings?.memory
              ),
            },
          ];

          // Process each metric
          metrics.forEach(({ metricType, metricSetting }) => {
            if (metricSetting?.enabled === 'true') {
              const [time, timeUnits] = getTimeAndUnitFromString(metricSetting.window_size);
              setMetricSettings(metricType, {
                isEnabled: true,
                currTopN: metricSetting.top_n_size ?? DEFAULT_TOP_N_SIZE,
                currWindowSize: time,
                currTimeUnit: timeUnits,
              });
            }
          });
          const groupBy = getMergedGroupBySettings(
            persistentSettings?.group_by,
            transientSettings?.group_by
          );
          if (groupBy) {
            setGroupBySettings({ groupBy });
          }
        } catch (error) {
          console.error('Failed to retrieve settings:', error);
        }
      } else {
        try {
          setMetricSettings(metric, {
            isEnabled: enabled,
            currTopN: newTopN,
            currWindowSize: newWindowSize,
            currTimeUnit: newTimeUnit,
          });
          setGroupBySettings({ groupBy: newGroupBy });
          await core.http.put('/api/update_settings', {
            query: {
              metric,
              enabled,
              top_n_size: newTopN,
              window_size: `${newWindowSize}${newTimeUnit === 'MINUTES' ? 'm' : 'h'}`,
              group_by: newGroupBy,
            },
          });
        } catch (error) {
          console.error('Failed to set settings:', error);
        }
      }
    },
    [core]
  );

  const onTimeChange = ({ start, end }: { start: string; end: string }) => {
    const usedRange = recentlyUsedRanges.filter(
      (range) => !(range.start === start && range.end === end)
    );
    usedRange.unshift({ start, end });
    setStart(start);
    setEnd(end);
    setRecentlyUsedRanges(usedRange.length > 10 ? usedRange.slice(0, 9) : usedRange);
    retrieveConfigInfo(true);
    retrieveQueries(start, end);
  };

  useEffect(() => {
    onTimeChange({ start: currStart, end: currEnd });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currStart, currEnd]);

  useEffect(() => {
    retrieveQueries(currStart, currEnd);
  }, [latencySettings, cpuSettings, memorySettings, currStart, currEnd, retrieveQueries]);

  return (
    <div style={{ padding: '35px 35px' }}>
      <Switch>
        <Route exact path="/query-details">
          {() => {
            return <QueryDetails core={core} depsStart={depsStart} />;
          }}
        </Route>
        <Route exact path="/query-group-details">
          {() => {
            return <QueryGroupDetails core={core} depsStart={depsStart} />;
          }}
        </Route>
        <Route exact path={QUERY_INSIGHTS}>
          <PageHeader
            coreStart={core}
            depsStart={depsStart}
            fallBackComponent={
              <>
                <EuiTitle size="l">
                  <h1>Query insights - Top N queries</h1>
                </EuiTitle>
                <EuiSpacer size="l" />
              </>
            }
          />
          <EuiTabs>{tabs.map(renderTab)}</EuiTabs>
          <EuiSpacer size="l" />
          <QueryInsights
            queries={queries}
            loading={loading}
            onTimeChange={onTimeChange}
            recentlyUsedRanges={recentlyUsedRanges}
            currStart={currStart}
            currEnd={currEnd}
            core={core}
          />
        </Route>
        <Route exact path={CONFIGURATION}>
          <PageHeader
            coreStart={core}
            depsStart={depsStart}
            fallBackComponent={
              <>
                <EuiTitle size="l">
                  <h1>Query insights - Configuration</h1>
                </EuiTitle>
                <EuiSpacer size="l" />
              </>
            }
          />

          <EuiTabs>{tabs.map(renderTab)}</EuiTabs>
          <EuiSpacer size="l" />
          <Configuration
            latencySettings={latencySettings}
            cpuSettings={cpuSettings}
            memorySettings={memorySettings}
            groupBySettings={groupBySettings}
            configInfo={retrieveConfigInfo}
            core={core}
          />
        </Route>
        <Redirect to={QUERY_INSIGHTS} />
      </Switch>
    </div>
  );
};

// eslint-disable-next-line import/no-default-export
export default TopNQueries;
