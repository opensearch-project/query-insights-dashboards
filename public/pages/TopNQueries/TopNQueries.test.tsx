/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { MemoryRouter } from 'react-router-dom';
import TopNQueries, { QUERY_INSIGHTS, CONFIGURATION } from './TopNQueries';
import { CoreStart } from 'opensearch-dashboards/public';

jest.mock('../QueryInsights/QueryInsights', () => () => <div>Mocked QueryInsights</div>);
jest.mock('../Configuration/Configuration', () => () => <div>Mocked Configuration</div>);
jest.mock('../QueryDetails/QueryDetails', () => () => <div>Mocked QueryDetails</div>);

const mockCore = ({
  http: {
    get: jest.fn(),
    put: jest.fn(),
  },
  uiSettings: {
    get: jest.fn().mockReturnValue(false),
  },
} as unknown) as CoreStart;

const setUpDefaultEnabledSettings = () => {
  const mockLatencyResponse = { response: { top_queries: [{ id: '1' }, { id: '2' }] } };
  const mockCpuResponse = { response: { top_queries: [{ id: '2' }, { id: '3' }] } };
  const mockMemoryResponse = { response: { top_queries: [{ id: '1' }] } };
  // Mock API responses for each metric
  (mockCore.http.get as jest.Mock).mockImplementation((endpoint) => {
    if (endpoint === '/api/top_queries/latency') return Promise.resolve(mockLatencyResponse);
    if (endpoint === '/api/top_queries/cpu') return Promise.resolve(mockCpuResponse);
    if (endpoint === '/api/top_queries/memory') return Promise.resolve(mockMemoryResponse);
    return Promise.resolve({ response: { top_queries: [] } });
  });
  // Mock API response for all metrics enabled
  const mockSettingsResponse = {
    response: {
      persistent: {
        search: {
          insights: {
            top_queries: {
              latency: { enabled: 'true', top_n_size: '10', window_size: '1h' },
              cpu: { enabled: 'true', top_n_size: '10', window_size: '1h' },
              memory: { enabled: 'true', top_n_size: '5', window_size: '30m' },
            },
          },
        },
      },
    },
  };
  (mockCore.http.get as jest.Mock).mockResolvedValueOnce(mockSettingsResponse);
};

const renderTopNQueries = (type: string) =>
  render(
    <MemoryRouter initialEntries={[type]}>
      <TopNQueries core={mockCore} depsStart={{ navigation: {} }} params={{} as any} />
    </MemoryRouter>
  );

describe('TopNQueries Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders and switches tabs correctly', () => {
    const container = renderTopNQueries(QUERY_INSIGHTS);

    // Check for Query Insights tab content
    expect(screen.getByText('Mocked QueryInsights')).toBeInTheDocument();
    expect(screen.getByText('Top N queries')).toBeInTheDocument();
    expect(screen.getByText('Configuration')).toBeInTheDocument();
    expect(container).toMatchSnapshot();

    // Switch to Configuration tab
    fireEvent.click(screen.getByText('Configuration'));
    expect(screen.getByText('Mocked Configuration')).toBeInTheDocument();
  });

  it('updates settings in retrieveConfigInfo based on API response', async () => {
    const mockSettingsResponse = {
      response: {
        persistent: {
          search: {
            insights: {
              top_queries: {
                latency: { enabled: 'true', top_n_size: '10', window_size: '1h' },
                cpu: { enabled: 'false' },
                memory: { enabled: 'true', top_n_size: '5', window_size: '30m' },
              },
            },
          },
        },
      },
    };
    (mockCore.http.get as jest.Mock).mockResolvedValueOnce(mockSettingsResponse);
    const container = renderTopNQueries(CONFIGURATION);
    await waitFor(() => {
      expect(mockCore.http.get).toHaveBeenCalledWith('/api/settings', {
        query: { dataSourceId: undefined },
      });
      expect(screen.getByText('Mocked Configuration')).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });

  it('fetches queries for all metrics in retrieveQueries', async () => {
    setUpDefaultEnabledSettings();
    const container = renderTopNQueries(QUERY_INSIGHTS);
    await waitFor(() => {
      // Verify each endpoint is called
      expect(mockCore.http.get).toHaveBeenCalledWith('/api/settings', expect.any(Object));
      expect(mockCore.http.get).toHaveBeenCalledWith(
        '/api/top_queries/latency',
        expect.any(Object)
      );
      expect(mockCore.http.get).toHaveBeenCalledWith('/api/top_queries/cpu', expect.any(Object));
      expect(mockCore.http.get).toHaveBeenCalledWith('/api/top_queries/memory', expect.any(Object));
      // Check that deduplicated queries would be displayed in QueryInsights
      expect(screen.getByText('Mocked QueryInsights')).toBeInTheDocument();

      expect(container).toMatchSnapshot();
    });
  });

  it('fetches queries for only enabled metrics in retrieveQueries', async () => {
    const mockResponse = { response: { top_queries: [{ id: '1' }, { id: '2' }] } };
    // Mock API responses for each metric
    (mockCore.http.get as jest.Mock).mockImplementation((endpoint) => {
      if (endpoint === '/api/top_queries/latency') return Promise.resolve(mockResponse);
      if (endpoint === '/api/top_queries/cpu') return Promise.resolve(mockResponse);
      if (endpoint === '/api/top_queries/memory') return Promise.resolve(mockResponse);
      return Promise.resolve({ response: { top_queries: [] } });
    });
    // Mock API response for only one metrics enabled
    const mockSettingsResponse = {
      response: {
        persistent: {
          search: {
            insights: {
              top_queries: {
                latency: { enabled: 'true', top_n_size: '10', window_size: '1h' },
                cpu: { enabled: 'false' },
                memory: { enabled: 'false' },
              },
            },
          },
        },
      },
    };
    (mockCore.http.get as jest.Mock).mockResolvedValueOnce(mockSettingsResponse);
    const container = renderTopNQueries(QUERY_INSIGHTS);
    await waitFor(() => {
      // Verify each endpoint is called
      expect(mockCore.http.get).toHaveBeenCalledWith('/api/settings', expect.any(Object));
      expect(mockCore.http.get).toHaveBeenCalledWith(
        '/api/top_queries/latency',
        expect.any(Object)
      );
      expect(mockCore.http.get).not.toHaveBeenCalledWith(
        '/api/top_queries/cpu',
        expect.any(Object)
      );
      expect(mockCore.http.get).not.toHaveBeenCalledWith(
        '/api/top_queries/memory',
        expect.any(Object)
      );
      // Check that deduplicated queries would be displayed in QueryInsights
      expect(screen.getByText('Mocked QueryInsights')).toBeInTheDocument();

      expect(container).toMatchSnapshot();
    });
  });

  it('updates time range and fetches data when time range changes', async () => {
    setUpDefaultEnabledSettings();
    (mockCore.http.get as jest.Mock).mockResolvedValueOnce({ response: { top_queries: [] } });
    // Render with initial time range
    const { rerender } = render(
      <MemoryRouter initialEntries={[QUERY_INSIGHTS]}>
        <TopNQueries
          core={mockCore}
          initialStart="now-1d"
          initialEnd="now"
          depsStart={{ navigation: {} }}
          params={{} as any}
        />
      </MemoryRouter>
    );
    // Mock a new response for the time range update
    (mockCore.http.get as jest.Mock).mockResolvedValueOnce({
      response: { top_queries: [{ id: 'newQuery' }] },
    });

    // Re-render with updated time range to simulate a change
    rerender(
      <MemoryRouter initialEntries={[QUERY_INSIGHTS]}>
        <TopNQueries
          core={mockCore}
          initialStart="now-7d"
          initialEnd="now"
          depsStart={{ navigation: {} }}
          params={{} as any}
        />
      </MemoryRouter>
    );
    // Verify that the component re-fetches data for the new time range
    await waitFor(() => {
      expect(mockCore.http.get).toHaveBeenCalledTimes(7);
      expect(mockCore.http.get).toHaveBeenCalledWith('/api/settings', expect.any(Object));
      expect(mockCore.http.get).toHaveBeenCalledWith(
        '/api/top_queries/latency',
        expect.any(Object)
      );
      expect(mockCore.http.get).toHaveBeenCalledWith('/api/top_queries/cpu', expect.any(Object));
      expect(mockCore.http.get).toHaveBeenCalledWith('/api/top_queries/memory', expect.any(Object));
    });
  });
});
