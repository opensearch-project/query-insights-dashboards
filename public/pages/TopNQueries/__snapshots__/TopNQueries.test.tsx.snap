// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopNQueries Component fetches queries for all metrics in retrieveQueries 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="padding: 35px 35px;"
      >
        <h1
          class="euiTitle euiTitle--large"
        >
          Query insights - Top N queries
        </h1>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div
          class="euiTabs"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="euiTab euiTab-isSelected"
            role="tab"
            type="button"
          >
            <span
              class="euiTab__content"
            >
              Top N queries
            </span>
          </button>
          <button
            aria-selected="false"
            class="euiTab"
            role="tab"
            type="button"
          >
            <span
              class="euiTab__content"
            >
              Configuration
            </span>
          </button>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div>
          Mocked QueryInsights
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="padding: 35px 35px;"
    >
      <h1
        class="euiTitle euiTitle--large"
      >
        Query insights - Top N queries
      </h1>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiTabs"
        role="tablist"
      >
        <button
          aria-selected="true"
          class="euiTab euiTab-isSelected"
          role="tab"
          type="button"
        >
          <span
            class="euiTab__content"
          >
            Top N queries
          </span>
        </button>
        <button
          aria-selected="false"
          class="euiTab"
          role="tab"
          type="button"
        >
          <span
            class="euiTab__content"
          >
            Configuration
          </span>
        </button>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div>
        Mocked QueryInsights
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TopNQueries Component fetches queries for only enabled metrics in retrieveQueries 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="padding: 35px 35px;"
      >
        <h1
          class="euiTitle euiTitle--large"
        >
          Query insights - Top N queries
        </h1>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div
          class="euiTabs"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="euiTab euiTab-isSelected"
            role="tab"
            type="button"
          >
            <span
              class="euiTab__content"
            >
              Top N queries
            </span>
          </button>
          <button
            aria-selected="false"
            class="euiTab"
            role="tab"
            type="button"
          >
            <span
              class="euiTab__content"
            >
              Configuration
            </span>
          </button>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div>
          Mocked QueryInsights
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="padding: 35px 35px;"
    >
      <h1
        class="euiTitle euiTitle--large"
      >
        Query insights - Top N queries
      </h1>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiTabs"
        role="tablist"
      >
        <button
          aria-selected="true"
          class="euiTab euiTab-isSelected"
          role="tab"
          type="button"
        >
          <span
            class="euiTab__content"
          >
            Top N queries
          </span>
        </button>
        <button
          aria-selected="false"
          class="euiTab"
          role="tab"
          type="button"
        >
          <span
            class="euiTab__content"
          >
            Configuration
          </span>
        </button>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div>
        Mocked QueryInsights
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TopNQueries Component renders and switches tabs correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="padding: 35px 35px;"
      >
        <h1
          class="euiTitle euiTitle--large"
        >
          Query insights - Top N queries
        </h1>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div
          class="euiTabs"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="euiTab euiTab-isSelected"
            role="tab"
            type="button"
          >
            <span
              class="euiTab__content"
            >
              Top N queries
            </span>
          </button>
          <button
            aria-selected="false"
            class="euiTab"
            role="tab"
            type="button"
          >
            <span
              class="euiTab__content"
            >
              Configuration
            </span>
          </button>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div>
          Mocked QueryInsights
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="padding: 35px 35px;"
    >
      <h1
        class="euiTitle euiTitle--large"
      >
        Query insights - Top N queries
      </h1>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiTabs"
        role="tablist"
      >
        <button
          aria-selected="true"
          class="euiTab euiTab-isSelected"
          role="tab"
          type="button"
        >
          <span
            class="euiTab__content"
          >
            Top N queries
          </span>
        </button>
        <button
          aria-selected="false"
          class="euiTab"
          role="tab"
          type="button"
        >
          <span
            class="euiTab__content"
          >
            Configuration
          </span>
        </button>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div>
        Mocked QueryInsights
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TopNQueries Component updates settings in retrieveConfigInfo based on API response 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="padding: 35px 35px;"
      >
        <h1
          class="euiTitle euiTitle--large"
        >
          Query insights - Configuration
        </h1>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div
          class="euiTabs"
          role="tablist"
        >
          <button
            aria-selected="false"
            class="euiTab"
            role="tab"
            type="button"
          >
            <span
              class="euiTab__content"
            >
              Top N queries
            </span>
          </button>
          <button
            aria-selected="true"
            class="euiTab euiTab-isSelected"
            role="tab"
            type="button"
          >
            <span
              class="euiTab__content"
            >
              Configuration
            </span>
          </button>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div>
          Mocked Configuration
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="padding: 35px 35px;"
    >
      <h1
        class="euiTitle euiTitle--large"
      >
        Query insights - Configuration
      </h1>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiTabs"
        role="tablist"
      >
        <button
          aria-selected="false"
          class="euiTab"
          role="tab"
          type="button"
        >
          <span
            class="euiTab__content"
          >
            Top N queries
          </span>
        </button>
        <button
          aria-selected="true"
          class="euiTab euiTab-isSelected"
          role="tab"
          type="button"
        >
          <span
            class="euiTab__content"
          >
            Configuration
          </span>
        </button>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div>
        Mocked Configuration
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
